### PR TITLE
AP_Motors: Add interpolated matrix backend

### PR DESCRIPTION
**Pull request by IamPete1**
This adds a motors back-end that allows dynamically changing the per-motor roll, pitch, yaw and throttle factors. This enables vehicles that change geometry. This was originally added for the transwing tilt-rotor aircraft. https://youtu.be/-QAy2u33BEc

This is only a 1D interpolation so can only cope with changing geometry with 1 DoF.

This currently loads everything from scripting, the interpolation is then done completely in C++ so scripting cannot go away and leave motors hanging. However we do still rely on scripting to update the interpolation point, so scripting going away could be bad still. 

If we were to directly load the factors in flight from scripting, it would be a very minimal C++ change and would allow as many degrees of freedom needed. However the script crashing would probably result in the vehicle crashing. 

I'm not really sure which is best, basically it boils down to how much we trust scripting.

### CODE REVIEW

**Comment by IamPete1** [comment]
Changed this so interpolation is done in scripting and loaded in flight. This allows much greater freedom but relies on the script working correctly.

File: libraries/AP_Motors/AP_MotorsMatrix_Scripting_Dynamic.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+#include "AP_MotorsMatrix_Scripting_Dynamic.h"
+#include <GCS_MAVLink/GCS.h>
+#include <AP_HAL/AP_HAL.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define debug_print 0
+
+// add a motor and give its testing order
+bool AP_MotorsMatrix_Scripting_Dynamic::add_motor(int8_t motor_num, uint8_t testing_order)
```
Comment (tridge) [inline libraries/AP_Motors/AP_MotorsMatrix_Scripting_Dynamic.cpp:28]: why signed motor_num?

File: libraries/AP_Motors/AP_MotorsMatrix_Scripting_Dynamic.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    }
+    if (motor_num >= 0 && motor_num < AP_MOTORS_MAX_NUM_MOTORS) {
+        _test_order[motor_num] = testing_order;
+        motor_enabled[motor_num] = true;
+        return true;
+    }
+    return false;
+}
+
+void AP_MotorsMatrix_Scripting_Dynamic::load_factors(factor_table &new_table)
```
Comment (tridge) [inline libraries/AP_Motors/AP_MotorsMatrix_Scripting_Dynamic.cpp:42]: should be const reference?

File: libraries/AP_Motors/AP_MotorsMatrix_Scripting_Dynamic.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            hal.console->printf("%i - Roll: %0.2f, Pitch %0.2f, Yaw: %0.2f, throttle %0.2f\n",i,_roll_factor[i],_pitch_factor[i],_yaw_factor[i],_throttle_factor[i]);
+        }
+    }
+#endif
+
+}
+
+bool AP_MotorsMatrix_Scripting_Dynamic::init(uint8_t expected_num_motors)
+{
+    // Check factors have been set
```
Comment (tridge) [inline libraries/AP_Motors/AP_MotorsMatrix_Scripting_Dynamic.cpp:65]: missing WITH_SEMAPHORE()

File: libraries/AP_Scripting/generator/description/bindings.desc [hunk 4]
```diff
@@ -372,3 +372,17 @@ singleton hal.gpio method pinMode void uint8_t 0 UINT8_MAX uint8_t 0 1
...
+
+include AP_Motors/AP_MotorsMatrix_Scripting_Dynamic.h depends APM_BUILD_TYPE(APM_BUILD_ArduPlane)||APM_BUILD_TYPE(APM_BUILD_ArduCopter)
+singleton AP_MotorsMatrix_Scripting_Dynamic depends APM_BUILD_TYPE(APM_BUILD_ArduPlane)||APM_BUILD_TYPE(APM_BUILD_ArduCopter)
+singleton AP_MotorsMatrix_Scripting_Dynamic alias Motors_dynamic
+singleton AP_MotorsMatrix_Scripting_Dynamic method init boolean uint8_t 0 (AP_MOTORS_MAX_NUM_MOTORS-1)
+singleton AP_MotorsMatrix_Scripting_Dynamic method add_motor void int8_t 0 (AP_MOTORS_MAX_NUM_MOTORS-1) uint8_t 0 (AP_MOTORS_MAX_NUM_MOTORS-1)
+singleton AP_MotorsMatrix_Scripting_Dynamic method load_factors void AP_MotorsMatrix_Scripting_Dynamic::factor_table
+
+userdata AP_MotorsMatrix_Scripting_Dynamic::factor_table depends APM_BUILD_TYPE(APM_BUILD_ArduPlane)||APM_BUILD_TYPE(APM_BUILD_ArduCopter)
+userdata AP_MotorsMatrix_Scripting_Dynamic::factor_table alias factor_table
```
Comment (tridge) [inline libraries/AP_Scripting/generator/description/bindings.desc:384]: alias to motor_factor_table would be better

**Review by tridge — CHANGES_REQUESTED** [review]

File: ArduCopter/Parameters.cpp [hunk 5]
```diff
@@ -870,7 +870,7 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
     // @Param: FRAME_CLASS
     // @DisplayName: Frame Class
     // @Description: Controls major frame class for multicopter component
-    // @Values: 0:Undefined, 1:Quad, 2:Hexa, 3:Octa, 4:OctaQuad, 5:Y6, 6:Heli, 7:Tri, 8:SingleCopter, 9:CoaxCopter, 10:BiCopter, 11:Heli_Dual, 12:DodecaHexa, 13:HeliQuad, 14:Deca, 15:Scripting Matrix, 16:6DoF Scripting
+    // @Values: 0:Undefined, 1:Quad, 2:Hexa, 3:Octa, 4:OctaQuad, 5:Y6, 6:Heli, 7:Tri, 8:SingleCopter, 9:CoaxCopter, 10:BiCopter, 11:Heli_Dual, 12:DodecaHexa, 13:HeliQuad, 14:Deca, 15:Scripting Matrix, 16:6DoF Scripting, 17:Dynamic scripting matrix
```
Comment (rmackay9) [inline ArduCopter/Parameters.cpp:873]: nitpick: "Dynamic Scripting Matrix"

File: libraries/AP_Motors/AP_MotorsMatrix_Scripting_Dynamic.cpp (hunk 1 above)
Comment (IamPete1) [reply libraries/AP_Motors/AP_MotorsMatrix_Scripting_Dynamic.cpp]: fixed, many of the other back ends have the same thing.

File: libraries/AP_Motors/AP_MotorsMatrix_Scripting_Dynamic.cpp (hunk 2 above)
Comment (IamPete1) [reply libraries/AP_Motors/AP_MotorsMatrix_Scripting_Dynamic.cpp]: yes, fixed

File: libraries/AP_Motors/AP_MotorsMatrix_Scripting_Dynamic.cpp (hunk 3 above)
Comment (IamPete1) [reply libraries/AP_Motors/AP_MotorsMatrix_Scripting_Dynamic.cpp:67]: added

File: libraries/AP_Scripting/generator/description/bindings.desc (hunk 4 above)
Comment (IamPete1) [reply libraries/AP_Scripting/generator/description/bindings.desc]: changed

File: ArduCopter/Parameters.cpp (hunk 5 above)
Comment (IamPete1) [reply ArduCopter/Parameters.cpp]: Updated.

**Comment by IamPete1** [comment]
I have fixed the ranges on the other scripting methods, testing order is 1 indexed and so is expected number of motors, so we have been limited to 11 motors not the full 12.

**Review by tridge — APPROVED** [review]

**Pull request merged**